src/random/draw.ts

```diff
@@ -15,6 +15,9 @@ import { random } from 'radashi'
  * // => [1, 2, 3]
  * ```
  */
+export function draw<T>(arr: readonly [T, ...T[]]): T
+export function draw<T>(array: readonly T[]): T | null
+
 export function draw<T>(array: readonly T[]): T | null {
```

Comment:
The line 19 and 21 have the same types. In my opinion, it is unnecessary since other functions with overload in the project do not do this duplication

Reply:
I agree in principle, but if I remove line 19, TS isn't happy:

```ts
const a = draw([])
// TS2345: Argument of type [] is not assignable to parameter of type readonly [unknown, ...unknown[]]
//   Source has 0 element(s) but target requires 1
```

My understanding of definition overloads in conjunction with implementation signature may be lacking here, but the compiler hints at an explanation:

> draw.ts(20, 17): The call would have succeeded against this implementation, but implementation signatures of overloads are not externally visible.

Reply:
Okay, [TS documentation](https://www.typescriptlang.org/docs/handbook/2/functions.html#overload-signatures-and-the-implementation-signature) is clear on this:

> The signature of the implementation is not visible from the outside. When writing an overloaded function, you should always have two or more signatures above the implementation of the function.

Reply:
One detail that I still can't wrap my head around: Given the signatures

```ts
export function draw<T>(arr: readonly [T, ...T[]]): T
export function draw<T>(array: readonly T[]): T | null
```

How is Typescript able to tell that `draw([])` results in `null`?
I can't see how the signatures imply this, but it seems to work.

Reply:
Perhaps the generic signature `draw<T>(array: readonly T[]): T | null` itself reduces to `draw(array: readonly never[]): null` when provided with `T = never`.